spanner/session.go

```diff
@@ -1107,6 +1208,79 @@ func (p *sessionPool) take(ctx context.Context) (*sessionHandle, error) {
 	}
 }
 
+// takeMultiplexed returns a cached session if there is available one; if there isn't
+// any, it tries to allocate a new one.
+func (p *sessionPool) takeMultiplexed(ctx context.Context) (*sessionHandle, error) {
+	if !p.enableMultiplexSession {
+		return p.take(ctx)
+	}
+	trace.TracePrintf(ctx, nil, "Acquiring a multiplexed session")
+	for {
+		var s *session
+
+		p.mu.Lock()
+		if !p.valid {
+			p.mu.Unlock()
+			return nil, errInvalidSessionPool
+		}
+		// use the multiplex session if it is available
+		if p.multiplexedSession != nil {
+			// Multiplexed session is available, get it.
+			s = p.multiplexedSession
+			trace.TracePrintf(ctx, map[string]interface{}{"sessionID": s.getID()},
+				"Acquired multiplexed session")
+		}
+		if s != nil {
+			p.mu.Unlock()
+			// From here, If healthcheck workers failed to
+			// schedule healthcheck for the session timely, do the check here.
+			// Because session check is still much cheaper than session
+			// creation, they should be reused as much as possible.
+			if !p.isHealthy(s) {
+				continue
+			}
+			p.incNumMultiplexedInUse(ctx)
+
+			return p.newSessionHandle(s), nil
+		}
+
+		// No session available. Start the creation of multiplexed session.
+		if err := p.getMultiplexedSession(ctx); err != nil {
+			p.mu.Unlock()
+			return nil, err
+		}
+
+		mayGetSession := p.mayGetMultiplexedSession
+		p.mu.Unlock()
+		trace.TracePrintf(ctx, nil, "Waiting for multiplexed session to become available")
+		select {
+		case <-ctx.Done():
+			trace.TracePrintf(ctx, nil, "Context done waiting for multiplexed session")
+			p.recordStat(ctx, GetSessionTimeoutsCount, 1, tag.Tag{Key: tagKeyIsMultiplexed, Value: "true"})
+			if p.otConfig != nil {
+				p.recordOTStat(ctx, p.otConfig.getSessionTimeoutsCount, 1, attributeKeyIsMultiplexed.String("true"))
+			}
+			p.mu.Lock()
+			p.mu.Unlock()
+			return nil, p.errGetSessionTimeout(ctx)
+		case <-mayGetSession:
+			p.mu.Lock()
+			if p.multiplexedSessionCreationError != nil {
```

Comment:
This property doesn't seem to be used anywhere else. That means that:
1. If the error is `Unimplemented`, then we fall back to using regular sessions.
2. If the error is for example `Database not found` or `Permission denied`, then we will continue to try to create a new multiplexed session every time that the method is called.

Is that intentional? Or do we want to return the `p.multiplexedSessionCreationError` in specific cases directly without trying to create a new one every time?

Reply:
We return the error below if it's not Unimplemented error